### PR TITLE
[geom] Correct trap safety and fitting behavior

### PR DESCRIPTION
**Pull request by dkonst13**
## It fixes:
- wrong outside safety in TGeoGtra::DistFromOutside
- wrong fitted box size in TGeoTrd1::GetFittingBox
- wrong fitted box size in TGeoTrd2::GetFittingBox

For TGeoGtra, the outside-distance code used the wrong safety mode. It used inside safety for a point that is outside the shape. This could return wrong safety values.

For TGeoTrd1 and TGeoTrd2, GetFittingBox could return a box that does not really fit inside the shape for some negative offsets.

## Checklist:

- [ ] tested changes locally
- [ ] updated the docs (if necessary)

This PR could be related to #20981

### CODE REVIEW

**Comment by dpiparo** [comment]
Thanks for this PR.
I would like to note that the format of the commit messages is not correct. All messages should start with `[geom]`
Moreover, I was wondering if a test for this bug fix can be provided

File: geom/geom/src/TGeoTrd2.cxx [hunk 1]
```diff
@@ -601,11 +601,11 @@ Int_t TGeoTrd2::GetFittingBox(const TGeoBBox *parambox, TGeoMatrix *mat, Double_
...
    Double_t z = origin[2] - dd[2];
-   dd[0] = dx0 - fx * z - origin[0];
-   dd[1] = dy0 - fy * z - origin[1];
+   dd[0] = dx0 - fx * z - TMath::Abs(origin[0]);
+   dd[1] = dy0 - fy * z - TMath::Abs(origin[1]);
    z = origin[2] + dd[2];
-   dd[0] = TMath::Min(dd[0], dx0 - fx * z - origin[0]);
-   dd[1] = TMath::Min(dd[1], dy0 - fy * z - origin[1]);
+   dd[0] = TMath::Min(dd[0], dx0 - fx * z - TMath::Abs(origin[0]));
+   dd[1] = TMath::Min(dd[1], dy0 - fy * z - TMath::Abs(origin[1]));
```
Comment (agheata) [inline geom/geom/src/TGeoTrd2.cxx:608]: The fix is partially correct; the method is supposed to compute the actual size of the daughter placed in this Trd with **either** or **both** dx, dy parameters negative, to fit this one. However, it will override a fixed positive child dimension. If only `parambox->GetDX() ` is negative, the dy of the box is also unconditionally changed to fit the parent Trd, while the contract is to stay untouched. So the fix should read like:
```c++
Bool_t runtimeX = (dd[0] < 0);
Bool_t runtimeY = (dd[1] < 0);
Double_t x = TMath::Abs(origin[0]);
Double_t y = TMath::Abs(origin[1]);
Double_t z = origin[2] - dd[2];
if (runtimeX)
   dd[0] = dx0 - fx * z - x;
if (runtimeY)
   dd[1] = dy0 - fy * z - y;
z = origin[2] + dd[2];
if (runtimeX)
   dd[0] = TMath::Min(dd[0], dx0 - fx * z - x);
if (runtimeY)
   dd[1] = TMath::Min(dd[1], dy0 - fy * z - y);
```

**Review by agheata — CHANGES_REQUESTED** [review]
Thanks a lot for the fix @dkonst13 ! These corner cases were indeed bugged and went unnoticed for years. Please check my change suggestion for TGeoTrd2. The fix in TGeoGtra is obvious, but the ones in TGeoTrd1/2 deserve a regression test. A natural place would be in geom/test, plus a ROOT_ADD_GTEST entry in geom/test/CMakeLists.txt. The test could look like:
```cpp
#include <gtest/gtest.h>

#include <TGeoBBox.h>
#include <TGeoMatrix.h>
#include <TGeoTrd1.h>
#include <TGeoTrd2.h>

TEST(Geometry, Trd1FittingBoxUsesNearestXFace)
{
   TGeoTrd1 mother(10., 10., 7., 5.);
   TGeoBBox child(-1., 1., 2.);
   TGeoTranslation placement(-8., 0., 0.);

   Double_t dx = 0.;
   Double_t dy = 0.;
   Double_t dz = 0.;
   EXPECT_EQ(mother.GetFittingBox(&child, &placement, dx, dy, dz), 0);
   EXPECT_DOUBLE_EQ(dx, 2.);
   EXPECT_DOUBLE_EQ(dy, 1.);
   EXPECT_DOUBLE_EQ(dz, 2.);
}

TEST(Geometry, Trd2FittingBoxUsesNearestFaces)
{
   TGeoTrd2 mother(10., 10., 8., 8., 5.);
   TGeoBBox child(-1., -1., 2.);
   TGeoTranslation placement(-8., -6., 0.);

   Double_t dx = 0.;
   Double_t dy = 0.;
   Double_t dz = 0.;
   EXPECT_EQ(mother.GetFittingBox(&child, &placement, dx, dy, dz), 0);
   EXPECT_DOUBLE_EQ(dx, 2.);
   EXPECT_DOUBLE_EQ(dy, 2.);
   EXPECT_DOUBLE_EQ(dz, 2.);
}

TEST(Geometry, Trd2FittingBoxPreservesFixedDimensions)
{
   TGeoTrd2 mother(10., 10., 8., 8., 5.);
   TGeoBBox child(1.5, -1., 2.);
   TGeoTranslation placement(4., -6., 0.);

   Double_t dx = 0.;
   Double_t dy = 0.;
   Double_t dz = 0.;
   EXPECT_EQ(mother.GetFittingBox(&child, &placement, dx, dy, dz), 0);
   EXPECT_DOUBLE_EQ(dx, 1.5);
   EXPECT_DOUBLE_EQ(dy, 2.);
   EXPECT_DOUBLE_EQ(dz, 2.);
}
```

**Comment by agheata** [comment]
@dkonst13 is there a follow-up to this?
